src/ServiceControl.Monitoring/Timings/TimingsStore.cs

```diff
@@ -0,0 +1,141 @@
+﻿namespace ServiceControl.Monitoring.Timings
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using Metrics.Raw;
+
+    public abstract class TimingsStore
+    {
+        ConcurrentDictionary<string, ConcurrentDictionary<DateTime, MeasurementInterval>> timings = 
+            new ConcurrentDictionary<string, ConcurrentDictionary<DateTime, MeasurementInterval>>();
+
+        public void Store(string endpointName, LongValueOccurrences message, DateTime now)
+        {
+            var endpointData = timings.GetOrAdd(endpointName, _ => new ConcurrentDictionary<DateTime, MeasurementInterval>());
+
+            for (var i = 0; i < message.Ticks.Length; i++)
+            {
+                var date = new DateTime(message.BaseTicks + message.Ticks[i], DateTimeKind.Utc);
+                var intervalId = date.RoundDownToNearest(IntervalSize);
+
+                endpointData.AddOrUpdate(
+                    intervalId,
+                    _ => new MeasurementInterval(1, message.Values[i]),
+                    (_, b) => b.Update(message.Values[i]));
+            }
+
+            if (endpointData.Count > 2 * NumberOfHistoricalIntervals)
```

Comment:
With the proposal above, cleaning oldest interval is just walking through the list and `Release`ing as many chunks as needed, starting from the head.